### ISSUE TITLE
Slow loading of instances (BB #522)

### ISSUE THREAD

**Issue by OregonAsari** [OregonCore/OregonCore#522]
This issue was migrated from bitbucket.
**Original Reporter:** 
**Original Date:** 03.03.2014 07:51:13 GMT+0000
**Original Priority:** minor
**Original Type:** bug
**Original State:** resolved
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/522

<hr>

When server loads and instances have cleanup and array for instance ids. It is very slow about 1instance / 3-5sec i think if you will have in server about 50 instances from people it may by slow for load of server.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 03.03.2014 11:42:56 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/522/#comment-8880212

<hr>

is this during startup of the server or during gameplay? if during gameplay are there significant lags ? Also what platform are you using ? we've noticed that windows is very slow at loading.. record is abou 44secs. however on my linux machine i get server up  in about 5secs.... this iz most probably due to jemalloc

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 04.03.2014 05:22:28 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/522/#comment-8901769

<hr>

I using linux debian.
I added to database 20 000 characters.  And tried to go 5 dungeons, after tried restart. Cleaning of instances was about 5sec/instance.

This function slowing start.
InstanceSaveMgr.cpp

```
#!c++

        if (*i != InstanceNumber)
        {
            // remap instance id
            WorldDatabase.PExecute("UPDATE creature_respawn SET instance = '%u' WHERE instance = '%u'", InstanceNumber, *i);
            WorldDatabase.PExecute("UPDATE gameobject_respawn SET instance = '%u' WHERE instance = '%u'", InstanceNumber, *i);
            CharacterDatabase.PExecute("UPDATE characters SET instance_id = '%u' WHERE instance_id = '%u'", InstanceNumber, *i);
            CharacterDatabase.PExecute("UPDATE corpse SET instance = '%u' WHERE instance = '%u'", InstanceNumber, *i);
            CharacterDatabase.PExecute("UPDATE character_instance SET instance = '%u' WHERE instance = '%u'", InstanceNumber, *i);
            CharacterDatabase.PExecute("UPDATE instance SET id = '%u' WHERE id = '%u'", InstanceNumber, *i);
            CharacterDatabase.PExecute("UPDATE group_instance SET instance = '%u' WHERE instance = '%u'", InstanceNumber, *i);
        }
```

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 04.03.2014 05:24:53 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/522/#comment-8901791

<hr>

On higher populated servers about 1000 players online i think they must have in database 100 000 players. And instances can be 100+ it must be very slow. I had quadcore. Fresh db is loaded after 3-6sec ;)

I tried comment these lines and start is fast on populated server :)

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 04.03.2014 14:44:15 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/522/#comment-8910772

<hr>

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 05.03.2014 15:01:32 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/522/#comment-8931962

<hr>

Speed up instance remapping, resolves #522
- Added missing index on table `characters`, don't forget to update.
- Moved from std::set to std::vector and asking database to sort ids for us
- Moved from regular queries to prepared stetements
- Database can be boosted quite a lot by user's configuration, but that is
  outside of our scope

→ <<cset 12a4330f13cb>>
